### ISSUE TITLE
Stream Scheduling

### ISSUE THREAD

**Issue by huitema** [private-octopus/picoquic#1295]
Picoquic employs a simple stream scheduling strategy:

* applications decide when to open a stream, and when to provide data to send on that stream (i.e., mark active API)
* if data is provided on several streams, the code will send data from the lowest numbered stream first.

In theory, applications who want to exercise fine-grain scheduling can do that by controlling the active/inactive mark for all the streams.  In practice that's a bit difficult: applications need to implement fine grain scheduling themselves, instead of merely posting a set of "pages", and they need to do so without introducing extra delays. For example, applications may want to:

* load a main page (html) and then load a series of images in parallel, using progressive rendering
* process multiple media stream in parallel
* make sure that a "control" stream is given higher priority than regular data

The priority scheme of [HTTP2](https://datatracker.ietf.org/doc/html/rfc7540) was designed for handling common web scenarios. They include:

* mark a stream as dependent on another stream (or dependent on the root for streams with no dependency). If stream A is dependent on stream B, data from B will be sent before A.
* associate a weight to a stream. If B and C are dependent of A, B with weight 20 and C with weight 80, C will be "polled" 4 times for each time B is polled.

The HTTP2 scheme was not carried over to HTTP3, maybe because the tradeoff between complexity and performance was not obvious.

There seems to be three key factors:

* handle dependencies, be able to say "send stream X only after stream Y". 
* handle parallelism, be able to say "send stream A and stream B in parallel". The streams can be marked as dependent on the same head stream or on the origin, and data can be picked from either one.
* keep some memory, so as to visit all available streams instead of just one.

There are two degraded modes: handle all streams in parallel; and, process all streams in order. The first one is good when the traffic is mostly a set of media streams. The second one is fine for RPC-like models, e.g., DoQ. All streams in parallel would address the issue #1259

**Comment by huitema** [comment]
Applications that want serial handling can easily achieve it using available API. The application knows when it has indicated or sent a FIN bit on a stream. It can at that point mark the next stream as "active". There will be no gap. This can also be used to implement the "dependent stream" functionality: mark all dependent streams as "inactive" until the FIN has been indicated for the "parent" stream, and mark them active at that point. So maybe there is no direct need to handle dependencies in the stack, if the application can do it by itself.

This leaves parallelism and memory. If several streams are active, a solution would be some kind of "last polled stream" list, a variant of an LRU list.

**Comment by huitema** [comment]
The solution is probably to maintain an ordered list of candidate streams, with some sorting by:

* priority, from 0=highest to UINT64_MAX = lowest.
* last_time_stream_was_polled

Streams would remain in this list as long as data for the stream is available, either because data is queued or because the stream is marked available. The list will be updated: 

* when data becomes available, insert the stream in the poll list.
* if priority changes, update the position of the stream in the poll list.
* always poll in order.
* after polling successfully, update the last time polled and the position of the stream in the list.

The main issue there would be what happens if the stream is in the list but could not be serviced, maybe because there are no stream credits, or because the application reneges on making data available. The code will move to the next stream in the list. This will only be a problem if there are many streams in this position. Maybe streams with no stream credits should be removed from the list, and reinserted when credit arrives. As for applications messing up the "available" state, that probably should not be over-optimized. Let the application solve it.

HTTP-3 has a concept of dependent stream, such as for example loading the main HTML page before loading images. Encoding that inside picoquic is rather complicated, as the whole tree of dependencies would have to be maintained and coordinated with scheduling. It seems simpler to let the application handle that, either by not providing data for dependent stream until the whole upper level stream has been processed, or maybe setting the dependent stream at a lower priority and updating that priority when the dependent stream is sent.

**Comment by LPardue** [comment]
FWIW We tried to port H2 priorities to H3, it was too complex. H2 priorities on their own were also too complex and got deprecated.

Extensible priorities is the replacement for both - https://datatracker.ietf.org/doc/draft-ietf-httpbis-priority/

**Comment by LPardue** [comment]
With quiche, we actually run a larger number space in the QUIC layer than urgency's 0 through 7. That gives us some wiggle room to inject other things (like control streams) and some future extensibility. 

For datagram, this is an area if debate in Webtransport right now.

I wrote up one approach that melded with the extensible priority scheme pretty well but it seemed to get a lukewarm response so far. Still might help you making your considerations https://datatracker.ietf.org/doc/html/draft-pardue-masque-dgram-priority-01

**Comment by LPardue** [comment]
Whoops, forgot the link to WebTransport https://github.com/w3c/webtransport/issues/62

**Comment by huitema** [comment]
The QUIC WG deliberately left the issues of datagram "stream" and identification to the application. However, nothing prevents senders from implementing multiple datagram "queues" (or some equivalent), assigning a different priority to each queue. Picoquic provides two ways to create datagram frames: either the application queues a whole datagram; or the application registers as a datagram sender and gets polled when there is space in a packet for a datagram frame.

I think the natural API is to have the application decorates the datagrams with a "priority", allowing the datagram streams to be scheduled just like a stream of the same priority. So "queue this N bytes as a datagram of priority P", or "announce readiness to send datagrams at priority P", and get polled when the time has come. This does not require transport level standardization. 

There is still a bit of fuzziness for interleaving datagrams and streams. If the application has multiple incremental streams at the same priority P, it will implement some kind of "round robin" between the streams. How does the datagrams play with that? Is it something like "one round for each stream, then one round for the datagrams", or "one round for the datagrams, then one round for one of the streams of same priority?"

**Comment by LPardue** [comment]
IMO at any priority level, if there is equal amount of data of each type available, then you probably want a configurable ratio 1:N i.e. a sliding scale between preferring datagrams or stream data. This will likely be a bit fuzzy due to datagram quantisation but probably can get close enough. 

1:1 is probably an ok default for some applications, but others might want to weigh more towards sending a particular type.

**Comment by huitema** [comment]
The current management has the advantage of being simple. There are three API: insert a stream in the output queue, remove a stream from the queue when it is finished, and find the next available stream. Finding the next stream is done by browsing the list in order of priority, checking each stream 1 by 1, and retaining the first one that has data to send and can send it:

- is has data queued, or is marked active, and it is not blocked by flow control;
- or, it needs to send the FIN bit, or a stream reset message, or a stop sending message.

The list is ordered once, and there is already some notion of priority -- high or low. To get the full priorities, we will need to manage a priority level (and possibly change it), which is conceptually simple: just insert at the right place in the list -- but that's an O(N) operation, instead of just "insert in front for high priority, at end for others". If we want to manage "round robin" scheduling, we will need to keep track of when data was last sent on the stream, and pick the stream with the lowest date for a specific priority level. This means doing a "remove/insert" operation after sending stream data. This is O(N) today, because of the double linked list. Implementing a splay would make that O(logN), which seems better.

We could try also to use "data available" as a way to order, so "find available stream" would only consider the streams that are actually useful. That means reordering the list of streams when data is added or exhausted, which is quite a bit of overhead.

**Comment by huitema** [comment]
Solved in PR #1404.

**Comment by huitema** [comment]
There is actually one last fix needed. When applications are mixing streams and datagrams, datagrams are sent before stream frames. This mean that, for example, a high stream video can starve stream frames. This is easier to solve by the picoquic stack than by the application. For example, count the number of successive occasions in which picoquic could have sent something but did not because datagrams took all the bandwidth, and refrain to poll for more datagrams if this happens more than N times in a row, with N being a parameter that the application can set.

**Comment by LPardue** [comment]
This is an area of active discussion right now over in WenTransport land. You finding things in tune with that, which is good sign it's a common problem. 

Here's a lengthy thread for the w3c API but it relates to protocols too https://github.com/w3c/webtransport/issues/62. 

One idea I've had is to try and align datagram prioritization into the extensible priorities stream. And then leave it up to implementations to decide how to do bandwidth share between datagrams and streams; see https://datatracker.ietf.org/doc/draft-pardue-masque-dgram-priority/

**Comment by huitema** [comment]
I read the discussions in web transport, and you draft, quickly. One specificity of picoquic is that it relies on just-in-time scheduling, rather than queues. The application receives a callback when picoquic is ready to send a packet, with the callback API providing the amount of bytes that would be available for a datagram frame. The application can then decide what to send in that packet, or nothing.

Barring protocol changes, picoquic as few options. Basically, just two:

* Provide an API so the application can see the priority of the next scheduled STREAM_DATA frame, so a well designed application will only schedule datagram frames that carry data more urgent than that.
* Provide an API so the stack sometimes reserves packets for sending stream frames, rather than offering the application the possibility of filling the packet with a datagram frame.

**Comment by huitema** [comment]
I added a "worst case" test in which Audio and Video are sent as data streams, and a misbehaving datagram source sends 10MB of data as fast as allowed. Without any kind of prioritization between datagra and streams, the results are very bad. In the tests, the audio delay grew well over a second.

audio latency | average | St. Dev.| Max
----------|--------|-----|-----------
AV        | 10.5ms | 1ms | 24ms
AV + data | 18ms | 7.5ms | 61ms
AV + dgram (raw) | >1s  | >2s | >3s
AV + dgram (1/1) | 53ms | 17ms | 63ms

Adding a "round robin" protection between datagrams and streams suppresses the worst effects of the misbehaving source. The delay increases not because of scheduling, but because with the high sending rate queues are building up in the network.

**Comment by clarkzjw** [comment]
> Solved in PR #1404.

After this PR, https://github.com/private-octopus/picoquic/pull/1404/, is it true that, if I just add `picoquic_set_default_priority(qclient, 1);` in `picoquicdemo` after `qclient` is created, multiple streams will be downloaded in parallel(round robin)?

The log still shows the second stream starts downloading only after the first stream ends. 
```
...
Testing scenario: </1080/BBB-I-1080p.mpd;/1080/BBB-I-1080p.seg0-L0.svc;>
Max stream id bidir remote before start = 0 (0)
Starting client connection. Version = 1, I-CID: d1816371f5e1b21b
Max stream id bidir remote after start = 2048 (2048)
Max stream id bidir remote after 0rtt = 2048 (2048)
Opening stream 0 to GET /1080/BBB-I-1080p.mpd
Waiting for packets.
Client port (AF=30): 52454.
Negotiated ALPN: h3
Almost ready!

Connection established. Version = 1, I-CID: d1816371f5e1b21b, verified: 1
Stream 0 ended after 88846 bytes
Opening stream 4 to GET /1080/BBB-I-1080p.seg0-L0.svc
Stream 4 ended after 1045181 bytes
All done, Closing the connection.
Received a request to close the connection.
The connection is closed!
...
```

So I guess extra modifications are still needed for multiple streams to download in parallel for `picoquicdemo` like mentioned in https://github.com/private-octopus/picoquic/issues/1259#issuecomment-990642746?

**Comment by huitema** [comment]
Oops, my examples are wrong. Sorry. Odd priority numbers (lowest bit set) are FIFO, even numbers are Round Robin. From "picoquic.h", see a list of examples:
```
 * - 0: system priority, e.g., the "settings" stream in HTTP3 (FIFO)
                                                               ^^^^ Should say round robin
 * - 1: high priority data (round robin)
                            ^^^^^^^^ That's an error, should say FIFO
 * - 2: real time audio (round robin)
 * - 4: real time video (round robin)
 * - 7: urgent data, such as CSS or JSON files (FIFO)
 * - 8: progressive data such as JPG files (round robin) 
 * - 9: web data (FIFO)
 * - 255: background data (FIFO)
 ```

**Comment by clarkzjw** [comment]
Actually, I also tried setting even numbers as the priority, e.g., 4. But both the picoquic console log and [qvis](https://qvis.quictools.info) show that the streams are downloaded sequentially. 

Attached are qlogs. (GitHub doesn't support *.qlog so *.txt is appended.)
[9c8835d4a09d9758.3d98.server.qlog.txt](https://github.com/private-octopus/picoquic/files/9650213/9c8835d4a09d9758.3d98.server.qlog.txt)
[9c8835d4a09d9758.client.qlog.txt](https://github.com/private-octopus/picoquic/files/9650214/9c8835d4a09d9758.client.qlog.txt)

**Comment by huitema** [comment]
Look at this part of the client's qlog:
```
[32164, "transport", "datagram_received", { "byte_length": 823}],
[32164, "transport", "packet_received", { "packet_type": "1RTT", "header": { "packet_size": 823, "packet_number": 66, "dcid": "
"355fefa075369894", "quic_bit": 0 }, "frames": [{ 
    "frame_type": "stream", "id": 0, "offset": 88203, "length": 668, "fin": true , "begins_with": "3830702e73656732"}, { 
    "frame_type": "padding"}]}],
[32263, "info", "message", { "message": "Preparing GET on stream 4 for /1080/BBB-I-1080p.init.svc"}],
[32282, "transport", "datagram_received", { "byte_length": 55}],
[32282, "transport", "packet_received", { "packet_type": "1RTT", "header": { "packet_size": 55, "packet_number": 67, "dcid": "355fefa075369894" }, "frames": [{ 
    "frame_type": "time_stamp", "time_stamp": 3290}, { 
    "frame_type": "ack", "ack_delay": 0, "acked_ranges": [[5, 13]]}, { 
    "frame_type": "padding"}]}],
[32282, "recovery", "metrics_updated", {"cwnd": 16674,"bytes_in_flight": 89,"smoothed_rtt": 8325}],
[32294, "transport", "spin_bit_updated", { "state": true }],
[32294, "transport", "packet_sent", { "packet_type": "1RTT", "header": { "packet_size": 167, "packet_number": 14, "dcid": "a93948129be5672a" }, "frames": [{ 
    "frame_type": "time_stamp", "time_stamp": 4036}, { 
    "frame_type": "ack", "ack_delay": 1, "acked_ranges": [[59, 67]]}, { 
    "frame_type": "stream", "id": 4, "offset": 0, "length": 58, "fin": true , "begins_with": "01380000d1d7511a"}, { 
    "frame_type": "padding"}]}],
```
The client receives the last byte and the fin flag of the first file, on stream 0, at T=32164. It schedules a GET request after that, at T=32263. That request is sent to the server on stream 4 at T=32294. That is, your client has a loop of:

* get file,
* receive the entire file
* get the next file

Of course that will be FIFO, whatever the scheduling. If you want parallel, you have to modify the application to open the second file without waiting for the first one.

**Comment by clarkzjw** [comment]
I see, thanks, I'll try to update that.

**Comment by clarkzjw** [comment]
Hi, I recently started to work on custom multipath scheduling algorithms. Currently, in picoquic, on the client side, 

1. Enters `picoquic_packet_loop`, `picoquic_select_ex` checks any incoming packets, if any, submit to `picoquic_incoming_packet_ex`.
2. In `picoquic_prepare_next_packet_ex` and `picoquic_prepare_packet_ex`, first select next path based on `availability` and `priority`(https://github.com/private-octopus/picoquic/blob/master/picoquic/sender.c#L4368-L4403), then, `if_index = cnx->path[path_id]->if_index_dest`, and finally, create packet and prepare segments based on current connection state. 

What I want to achieve is to add a path preference option for streams on the application level. For example,

1. The client would like to download a series of files with different priorities in a multipath environment(e.g., two paths with different speeds and latencies). And there could be dependencies between these files.
2. The client creates a stream to download each file. 
3. In order to utilize two paths, the client wants to schedule the downloading of higher priority files to the faster path, i.e., the client set a preference to schedule specific streams to specific paths.

Sure, the above context assumes these two paths are both available and ready and don't suffer from serious packet losses. If during downloading the stream, the preferred path becomes unavailable, then it falls back to any other available paths. 

Or, the application could set different stream priorities and schedule higher priority streams to faster paths. 

I'm still thinking about how to design and implement such ideas, are there any major flaws or is there a better way to achieve such goals(let the application decide specific streams go to specific paths)?

**Comment by clarkzjw** [comment]
@huitema Any suggestions?

**Comment by huitema** [comment]
Sorry, I am currently traveling, and there are limits to what I can do.

What you are looking for is setting some affinity between streams and paths, and using that as part of the stream scheduling algorithm. Maybe start with a spec, and see how that could be implemented. And tested.

**Comment by huitema** [comment]
Let's close this issue, since stream scheduling is fixed and we have a separate issue for stream and path affinity.
